flink-formats/flink-csv/src/main/java/org/apache/flink/formats/csv/CsvRowSerializationSchema.java

```diff
@@ -0,0 +1,234 @@
+import java.sql.Date;
+import java.sql.Time;
+import java.sql.Timestamp;
+
+/**
+ * Serialization schema that serializes an object of Flink types into a CSV bytes.
+ *
+ * <p>Serializes the input row into a {@link ObjectNode} and
+ * converts it into <code>byte[]</code>.
+ *
+ * <p>Result <code>byte[]</code> messages can be deserialized using {@link CsvRowDeserializationSchema}.
+ */
+@PublicEvolving
+public class CsvRowSerializationSchema implements SerializationSchema<Row> {
+
+	/** Schema describing the input csv data. */
+	private CsvSchema csvSchema;
+
+	/** Type information describing the input csv data. */
+	private TypeInformation<Row> rowTypeInfo;
+
+	/** CsvMapper used to write {@link JsonNode} into bytes. */
+	private CsvMapper csvMapper = new CsvMapper();
+
+	/** Reusable object node. */
+	private ObjectNode root;
+
+	/** Charset for byte[]. */
+	private String charset = "UTF-8";
+
+	/**
+	 * Create a {@link CsvRowSerializationSchema} with given {@link TypeInformation}.
+	 * @param rowTypeInfo type information used to create schem.
+	 */
+	CsvRowSerializationSchema(TypeInformation<Row> rowTypeInfo) {
+		Preconditions.checkNotNull(rowTypeInfo, "rowTypeInfo must not be null !");
+		this.rowTypeInfo = rowTypeInfo;
+		this.csvSchema = CsvRowSchemaConverter.rowTypeToCsvSchema((RowTypeInfo) rowTypeInfo);
+	}
+
+	@Override
+	public byte[] serialize(Row row) {
+		if (root == null) {
+			root = csvMapper.createObjectNode();
+		}
+		try {
+			convertRow(root, row, (RowTypeInfo) rowTypeInfo);
+			return csvMapper.writer(csvSchema).writeValueAsBytes(root);
+		} catch (JsonProcessingException e) {
+			throw new RuntimeException("Could not serialize row '" + row + "'. " +
+				"Make sure that the schema matches the input.", e);
+		}
+	}
+
+	private void convertRow(ObjectNode reuse, Row row, RowTypeInfo rowTypeInfo) {
+		if (reuse == null) {
+			reuse = csvMapper.createObjectNode();
+		}
+		if (row.getArity() != rowTypeInfo.getFieldNames().length) {
+			throw new IllegalStateException(String.format(
+				"Number of elements in the row '%s' is different from number of field names: %d",
+				row, rowTypeInfo.getFieldNames().length));
+		}
+		TypeInformation[] types = rowTypeInfo.getFieldTypes();
+		String[] fields = rowTypeInfo.getFieldNames();
+		for (int i = 0; i < types.length; i++) {
+			String columnName = fields[i];
+			Object obj = row.getField(i);
+			reuse.set(columnName, convert(reuse, obj, types[i], false));
+		}
+	}
+
+	/**
+	 * Converts an object to a JsonNode.
+	 * @param container {@link ContainerNode} that creates {@link JsonNode}.
+	 * @param obj Object that used to {@link JsonNode}.
+	 * @param info Type infomation that decides the type of {@link JsonNode}.
+	 * @param nested variable that indicates whether the obj is in a nested structure
+	 *               like a string in an array.
+	 * @return result after converting.
+	 */
+	private JsonNode convert(ContainerNode<?> container, Object obj, TypeInformation info, Boolean nested) {
```

Comment:
This nested flag means that only one level of nesting is supported right?

Reply:
yes.